pycontrails/core/flight.py

```diff
@@ -1626,3 +1683,69 @@ def segment_rocd(
     out[-1] = np.nan
 
     return out
+
+
+def fit_altitude(
+    elapsed_time: npt.NDArray[np.float_],
+    altitude_ft: npt.NDArray[np.float_],
+    max_segments: int = 30,
+    pop: int = 3,
+    r2_target: float = 0.999,
+    max_cruise_rocd: float = 10,
+    sg_window: int = 7,
+    sg_polyorder: int = 1,
+) -> npt.NDArray[np.float_]:
+    """Use piecewise linear fitting to smooth a flight profile.
+
+    Fit a flight profile to a series of line segments. Segments that have a
+    small rocd will be set to have a slope of zero and snapped to the
+    nearest thousand foot level.  A Savitzky-Golay filter will then be
+    applied to the profile to smooth the climbs and descents.  This filter
+    works best for high frequency flight data, sampled at a 1-3 second
+    sampling period.
+
+    Parameters
+    ----------
+    elapsed_time: npt.NDArray[np.float_]
+        Cumulative time of flight between waypoints, [:math:`s`]
+    altitude_ft: npt.NDArray[np.float_]
+        Altitude of each waypoint, [:math:`ft`
+    max_segments: int, optional
+        The maximum number of line segements to fit to the flight profile.
+    pop: int, optional
+        Population parameter used for the stocastic optimization routine
+        used to fit the flight profile.
+    r2_target: float, optional
+        Target r^2 value for solver. Solver will continue to add line
+        segments until the resulting r^2 value is greater than this.
+    max_cruise_rocd: float, optional
+        The maximum ROCD for a segment that will be forced to a slope of
+        zero, [:math:`ft s^{-1}`]
+    sg_window: int, optional
+        Parameter for :func:`scipy.signal.savgol_filter`
+    sg_polyorder: int, optional
+        Parameter for :func:`scipy.signal.savgol_filter`
+
+    Returns
+    -------
+    npt.NDArray[np.float_]
+        Smoothed flight altitudes
+    """
+    for i in range(1, max_segments):
+        m2 = pwlf.PiecewiseLinFit(elapsed_time, altitude_ft)
+        r = m2.fitfast(i, pop)
+        r2 = m2.r_squared()
+        if r2 > r2_target:
+            break
+
+    mask = abs(m2.slopes) < max_cruise_rocd / 60.0
+    bounds = r[:-1][mask], r[1:][mask]
+    lvl = np.round(m2.intercepts[mask] / 1000) * 1000
+    for i in range(len(bounds[0])):
+        altitude_ft[
+            np.bitwise_and(elapsed_time >= float(bounds[0][i]), elapsed_time <= float(bounds[1][i]))
+        ] = lvl[i]
```

Comment:
Can this be vectorized? I may not be understanding it ... would something like this work:

```
filt = (elapsed_time >= bounds[0]) & (elapsed_time <= bounds[1])
altitude_ft[filt] = lvl[filt]
```
?

Reply:
That doesn't work, but this does, not sure if it's much better:

```
  bounds = r[:-1][mask], r[1:][mask]
  lvl = np.round(m2.intercepts[mask], -3)
  time_stack = np.repeat(elapsed_time[:,np.newaxis],lvl.size,axis=1)
  filt = (time_stack >= bounds[0]) & (time_stack <= bounds[1])
  for i in range(lvl.size):
      altitude_ft[filt[:,i]] = lvl[i]
```